### DISCUSSION TITLE
Enabling/Disabling preview features at startup with Keycloak.X

### DISCUSSION THREAD

**Discussion by danielFesenmeyer** [keycloak/keycloak#10986] [Archive]
Hi,

this issue is related to discussion [10886](https://github.com/keycloak/keycloak/discussions/10886).

We are currently migrating to Keycloak.X and we use the following system properties to enable preview features at runtime (startup):
```
          -Dkeycloak.profile.feature.admin_fine_grained_authz=enabled \
          -Dkeycloak.profile.feature.declarative_user_profile=enabled \
          -Dkeycloak.profile.feature.token_exchange=enabled
```

We don't set all of those system properties for all of our customers. Instead, each customer gets a subset of those preview features.

We still use the system-property approach documented for the Wildfly distribution, and it works fine with Quarkus - just because we don't use any feature which depends on an [EnvironmentDependentProviderFactory](https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/provider/EnvironmentDependentProviderFactory.java), such as scripts.

The question is whether this **system-property approach will be still supported, when Wildfly support is dropped?** Otherwise, we would have to switch to the approach documented [here](https://www.keycloak.org/server/features), using the --features build option. But this would lead to the topic described in [10886](https://github.com/keycloak/keycloak/discussions/10886):  we have to either deal with slow startup due to autobuild, build separate images per customer or implement a workaround.

**Comment by thomasdarimont** [comment]
There is also the option of `profile.properties` which still seems to be supported in keycloak.x (https://github.com/keycloak/keycloak/blob/8ee7ae24deaad3dbcf20859b69fa7c4709d65515/common/src/main/java/org/keycloak/common/Profile.java#L241)

**Comment by danielFesenmeyer** [comment]
When using the old system-property approach, for several features (at least those depending on  [EnvironmentDependentProviderFactory](https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/provider/EnvironmentDependentProviderFactory.java), there is an inconsistency in the Admin UI: Features are displayed as active, but the required providers are not there.

For example, if I provide a script named my-script-authenticator.js and enable the `scripts` feature at build time via `--features=scripts`, everything is fine: On the Server Info page _/admin/master/console/#/server-info_, `SCRIPTS` is not contained in the _Disabled Features_ section. And there is the expected provider on the providers page _/admin/master/console/#/server-info/providers_: SPI: authenticator, Provider: script-my-script-authenticator.js. I can also add it to an authentication flow.

But if I don't enable the feature at build time and use the system property `-Dkeycloak.profile.feature.scripts=enabled` at runtime, the provider script-my-script-authenticator.js is no longer shown on the providers view and can no longer be used in an authentication flow. But the Server Info page still does not show the `SCRIPTS` feature in the _Disabled Features_ section, whereas it is actually disabled.

**Reply by Arulaln-AR** [reply]
@danielFesenmeyer ,

Need help in enabling the script feature. I have mapped the script feature using the environment variable. Attached is the complete dockerfile. Built an image based on that and still under Authentication - > Add Execution step, I don't see an script option to create Javascript.

Do I need to add the javascript as well under the /opt/keycloak/providers/ during the build stage itself.
[Dockerfile.txt](https://github.com/keycloak/keycloak/files/9234910/Dockerfile.txt)

**Reply by danielFesenmeyer** [reply]
> Do I need to add the javascript as well under the /opt/keycloak/providers/ during the build stage itself.

That's exactly what I did - at least it worked with kc 17.0.0.

Actually you need to provide a jar file in the providers directory - it is described here and in previous sections: https://www.keycloak.org/docs/latest/server_development/#deploy-the-script-jar

**Reply by Arulaln-AR** [reply]
@danielFesenmeyer ,

Thanks Daniel, did those steps and can see my authentication js under Add Execution step. They made us to create extra jar file and push it to the providers folder for security purposes it seems.

**Comment by Arulaln-AR** [comment]
Hi Daniel,

If you have those details like the steps, could you please share it with
me? It would help me in trying for version 18.

On Mon, 1 Aug 2022 at 21:40, Daniel Fesenmeyer ***@***.***>
wrote:

> Do I need to add the javascript as well under the /opt/keycloak/providers/
> during the build stage itself.
>
> That's exactly what I did - at least it worked with kc 17.0.0.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keycloak/keycloak/discussions/10986#discussioncomment-3300195>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AP4YZ5I3RPBQTP6S3FTJ7E3VW7ZIPANCNFSM5R6XXBXA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Regards,
Arulaln A R
+91-8754438576

**Reply by Arulaln-AR** [reply]
Got the link, let me try to do the same and get back. Thanks for your help, really appreciate it.
